### PR TITLE
OpenRTB Prebid Cache support

### LINKED ISSUE #216: OpenRTB support: Bid cache
The `/openrtb2/auction` endpoint needs to support sending bids to the cache.

Right now, I don't think we have a good enough sense of the cache's use-cases to build a quality feature here. This is the fundamental tradeoff:

- Bids cached unnecessarily waste computing resources of the PBS host.
- Bids _not_ cached limits the publisher, to varying extends depending on the context.

If the cache behavior needs to be "chosen", because different behaviors have equal merit, then we have two options:

- If the choice lies with the publisher, then we can add to the OpenRTB `request.ext`
- If the choice lies with the host, we can add to the PBS Config

However... I also want to note the following, which we should take very seriously:

- Each "choice" adds code complexity, which increases the time of all future development.
- Choices are easy to add, but very hard to remove. Users will grow to depend on them.

To grow this project smoothly, our goal should be to identify the fewest "choices" _necessary_ to encourage adoption on both sides.
 
------------------------------

These are things to consider:

- The Prebid.JS philosophy is: "the publisher ultimately decides which bid wins." If PBS caches the bids, Prebid auctions will have overall better performance. However, it _is_ possible for JS expose an API which caches only the bid which the publisher chooses.

- The prebid-mobile philosophy is: "App code updates are expensive. It's better if the Server does _everything_, including picking the winning bid." As far as I can tell, there's no room for negotiation here.

- @DucChau mentioned that Rubicon wanted an option to cache _just_ the vast XML. Currently PBS caches a JSON blob with other info too.

- There's some ongoing debate about the cache's usefulness for certain types of bids in #199.

--------------------------------

These are some deficiencies in the `/auction` endpoint's cache behavior. Ideally, we can do it better in `openrtb2/auction`.

- Bids aren't cached until the auction has completed. So if _any_ bidder in the auction times out, then the cache calls do too. This basically wastes the auction on mobile, since it requires the bid to be cached.
- @DucChau noted that `/auction` doesn't let you cache _just_ the vast XML. It requires the JSON blob. If this is a valid use-case, then the `/openrtb2/auction` API should to support this too.

@bretg and @mjacobsonny might want to discuss this. At the moment, I don't think it's fleshed out well enough for much useful engineering work to be done.

### PR DESCRIPTION
**Pull request by dbemiller**
This fixes #216 and #199

The code is... pretty ugly, and a little embarrassing.

Normally I'd like to clean this up... but Prebid Mobile is being blocked by OpenRTB support in PBS, and this isn't the only thing left to do, so... I've gotta scope it somewhere.

### CODE REVIEW

**Comment by dbemiller** [comment]
@ndhimehta please send this to someone on rubicon as the second reviewer.

**Comment by ndhimehta** [comment]
@dbemiller Review scheduled for WO - 1/15.

File: exchange/auction.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+		if !ok || cpm > bestSoFar.Price {
+			bidMap[name] = bid
+		}
+	} else {
+		auction.winningBidsByBidder[bid.ImpID] = make(map[openrtb_ext.BidderName]*openrtb.Bid)
+		auction.winningBidsByBidder[bid.ImpID][name] = bid
+	}
+}
+
+func (auction *auction) numImps() int {
```
Comment (hhhjort) [inline exchange/auction.go]: Is this function supposed to have a purpose outside of the one test case?

Comment (dbemiller) [reply exchange/auction.go]: It makes `auction` nil-safe. The idea was to keep the "do we need to track winning bids?" logic in one place. If not, the `exchange` can initialize the variable to a `nil`. If so, the exchange can initialize a real `auction` object.

Then the rest of the `exchange` logic can call these functions _as if_ it needed to track the winning bids without much performance impact.

Comment (dbemiller) [reply exchange/auction.go]: Wait I'm sorry.... no, you're right. It's not called anywhere real.

I'll delete it.

File: exchange/cache.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	"encoding/json"
+	"github.com/golang/glog"
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"github.com/prebid/prebid-server/pbs/buckets"
+	"github.com/prebid/prebid-server/prebid_cache_client"
+	"strings"
+)
+
+// cacheBids stores the given Bids in prebid cache, and saves the generated UUIDs in the auction.
```
[the remark this answers is not in the document]
Comment (dbemiller) [reply exchange/cache.go]: agreed, thanks. I updated the comments to match.

**Comment by dbemiller** [comment]
One noteworthy change in the last few updates: @anwzhang pointed out that legacy mobile used `hb_cache_id` rather than `hb_uuid`. If we change this name, publishers have to update all their creatives in DFP.

So... I renamed the param to `hb_cache_id`, for legacy compatibility.

File: exchange/auction.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+		}
+	} else {
+		auction.winningBidsByBidder[bid.ImpID] = make(map[openrtb_ext.BidderName]*openrtb.Bid)
+		auction.winningBidsByBidder[bid.ImpID][name] = bid
+	}
+}
+
+func (auction *auction) cacheId(bid *openrtb.Bid) (id string, exists bool) {
+	id, exists = auction.cachedBids[bid]
+	return
```
Comment (sandraleon) [inline exchange/auction.go]: this returns values for id and exists?

File: exchange/auction.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+// forEachBestBid runs the callback function on every bid which is the highest one for each Bidder on each Imp.
+func (auction *auction) forEachBestBid(callback func(impID string, bidder openrtb_ext.BidderName, bid *openrtb.Bid, winner bool)) {
+	if auction == nil {
```
Comment (sandraleon) [inline exchange/auction.go]: is it necessary to check this? can auction be nil if it was able to call this function?

File: exchange/exchange.go [hunk 5]
```diff
@@ -41,11 +44,13 @@ type bidResponseWrapper struct {
...
 }
 
-func NewExchange(client *http.Client, cfg *config.Configuration) Exchange {
+func NewExchange(client *http.Client, cache prebid_cache_client.Client, cfg *config.Configuration) Exchange {
 	e := new(exchange)
 
 	e.adapterMap = newAdapterMap(client, cfg)
 	e.adapters = make([]openrtb_ext.BidderName, 0, len(e.adapterMap))
+	e.cache = cache
+	e.cacheTime = time.Duration(cfg.CacheURL.ExpectedTimeMillis) * time.Millisecond
```
Comment (sandraleon) [inline exchange/exchange.go]: if this is not set in the config, cache time will be zero?

File: exchange/auction.go (hunk 3 above)
Comment (dbemiller) [reply exchange/auction.go]: yes: https://tour.golang.org/basics/7

File: exchange/exchange.go (hunk 5 above)
Comment (dbemiller) [reply exchange/exchange.go]: correct. Do you think that makes sense as a default, or would it be better to have some time here?

Per [the note here](https://github.com/prebid/prebid-server/pull/276/files#diff-3baf47c64847a8fb8aaa8cc2e088513bR131), I hope this doesn't become a long-term solution in any case.

File: exchange/auction.go (hunk 4 above)
Comment (dbemiller) [reply exchange/auction.go]: no, you're right, it's not necessary. `auction` is non-nil if `targeting` is non-nil... and the only call to this function is after an `if targeting == nil { return; }` block.

I removed it.

File: exchange/exchange.go (hunk 5 above)
Comment (sandraleon) [reply exchange/exchange.go]: seems fine to me, just wanted to make sure that was what we wanted.

**Pull request merged**